src/aiodynamo/expressions.py

```diff
@@ -0,0 +1,537 @@
+            return UpdateExpression(remove={self})
+        return UpdateExpression(set_updates={self: Value(value)})
+
+    def set_if_not_exists(self, value: Any) -> UpdateExpression:
+        if isinstance(value, (bytes, str)) and not value:
+            return UpdateExpression()
+        return UpdateExpression(set_updates={self: IfNotExists(value)})
+
+    def change(self, diff: Numeric) -> UpdateExpression:
+        return UpdateExpression(set_updates={self: Modify(diff)})
+
+    def append(self, value: List[Any]) -> UpdateExpression:
+        return UpdateExpression(set_updates={self: Append(list(value))})
+
+    def remove(self) -> UpdateExpression:
+        return UpdateExpression(remove={self})
+
+    def add(self, value: Addable) -> UpdateExpression:
+        return UpdateExpression(add={self: value})
+
+    def delete(self, value: Set[Any]) -> UpdateExpression:
+        return UpdateExpression(delete={self: value})
+
+
+class KeyCondition(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def encode(self, params: Parameters) -> str:
+        pass
+
+
+@dataclass(frozen=True)
+class HashKey(KeyCondition):
+    name: str
+    value: Any
+
+    def encode(self, params: Parameters) -> str:
+        return f"{params.encode_path(self.name)} = {params.encode_value(self.value)}"
+
+    def __and__(self, other: Condition) -> KeyCondition:
+        return HashAndRangeKeyCondition(self, other)
+
+
+@dataclass(frozen=True)
+class RangeKey:
+    name: str
+
+    def begins_with(self, substr: str) -> Condition:
+        return BeginsWith(F(self.name), substr)
+
+    def between(self, low: Any, high: Any) -> Condition:
+        return Between(F(self.name), low, high)
+
+    def contains(
+        self, value: Union[str, bytes, int, float, decimal.Decimal]
+    ) -> Condition:
+        return Contains(F(self.name), value)
+
+    def size(self) -> Size:
+        return Size(F(self.name))
+
+    def is_in(self, values: Sequence[Any]) -> Condition:
+        return In(F(self.name), values)
+
+    def gt(self, other: Any) -> Condition:
+        return Comparison(F(self.name), ">", other)
+
+    def gte(self, other: Any) -> Condition:
+        return Comparison(F(self.name), ">=", other)
+
+    def lt(self, other: Any) -> Condition:
+        return Comparison(F(self.name), "<", other)
+
+    def lte(self, other: Any) -> Condition:
+        return Comparison(F(self.name), "<=", other)
+
+    def equals(self, other: Any) -> Condition:
+        return Comparison(F(self.name), "=", other)
+
+    def not_equals(self, other: Any) -> Condition:
+        return Comparison(F(self.name), "<>", other)
+
+
+class Parameters:
+    def __init__(self):
+        self.names: Dict[str, str] = {}
+        self.values: Dict[str, Dict[str, Any]] = {}
+        self.names_gen: Iterator[int] = count()
+        self.values_gen: Iterator[int] = count()
+        self.names_cache: _ParametersCache = {}
+        self.values_cache: _ParametersCache = {}
```

Comment:
Oh, I was so confused as to why `Parameters` were always created without arguments...
Now I see that it's stateful object and that it's passed around with the idea that information is recorded in it... 
I don't know how this could be done better, but it's confusing 🤔

Reply:
It's an internal API and users should never use it directly, so I think it's okay if it's not the best API. To prevent name clashes, all names and values get "placeholderified" by aiodynamo and this keeps track of all of them per API call. It's not the prettiest, I agree, but it get's the job done...